libraries/AC_Fence/AC_Fence.cpp

```diff
@@ -550,6 +573,74 @@ bool AC_Fence::pre_arm_check(char *failure_msg, const uint8_t failure_msg_len) c
     return true;
 }
 
+/*
+  get our altitude in the FENCE_ALT_TYPE frame
+  return false if not available
+ */
+bool AC_Fence::get_alt_U_in_frame(float &alt, Location::AltFrame alt_frame) const
+{
+    const auto &ahrs = AP::ahrs();
+
+    ahrs.get_relative_position_D_home(alt);
+
+    // translate to +ve up
+    alt = -alt;
+
+    /*
+      convert to desired altitude frame
+     */
+    switch (alt_frame) {
+    case Location::AltFrame::ABOVE_HOME:
+        break;
+    case Location::AltFrame::ABOVE_ORIGIN:
+        // try and get it relative origin
+        if (!ahrs.get_relative_position_D_origin_float(alt)) {
+            // treat as a breach if origin not known
+            return false;
+        }
+        alt = -alt;
+        break;
+    case Location::AltFrame::ABSOLUTE: {
+        float home_alt;
+        if (!ahrs.get_home().get_alt_m(Location::AltFrame::ABSOLUTE, home_alt)) {
+            // treat not known home alt as a breach
+            return false;
+        }
+        alt += home_alt;
+        break;
+    }
+    case Location::AltFrame::ABOVE_TERRAIN: {
+#if AP_TERRAIN_AVAILABLE
+        // get height above ground level with extrapolation. We extrapolate to prevent
+        // a fence breach due to a momentary cache miss in the terrain library
+        auto *terrain = AP::terrain();
+        if (terrain == nullptr || !terrain->height_above_terrain(alt, true)) {
+            // treat not known home alt as a breach
+            return false;
+        }
+        break;
+#else
+        // no terrain available
+        return false;
+#endif
+    }
+    }
+    return true;
+}
+
+// update safe alt min
```

Comment:
```suggestion
// update safe alt min - home *must* be set before this function is called
```

(and in the header file)

... or assert it is set in SITL somehow.  Or both.